src/potentials.cpp

```diff
@@ -504,92 +504,171 @@ void FunctorPotential::from_json(const json &j) {
-                    else
-                        break;
-                }
+    faunus_logger->trace("Pair potential spline tolerance = {} kT", js.value("utol", 1e-5));
 
-                assert(rmin2 < rmax2);
-
-                KnotData knotdata = spline.generate(
-                    [&](double r2) {
-                        return this->umatrix(i, k)(a, b, r2, {0, 0, 0});
-                    },
-                    rmin2, rmax2);
-
-                // assert if potential is negative for r<rmin
-                if (spline.eval(knotdata, knotdata.rmin2 + dr) < 0) {
-                    assert(hardsphere == false && "`hardsphere` is set, but potential is negative for r<rmin");
-                    knotdata.isNegativeBelowRmin = true;
-                } else
-                    matrix_of_knots.set(i, k, knotdata);
-
-                faunus_logger->debug("Potential for {}-{} splined with {} knot(s)", atoms[i].name, atoms[k].name,
-                                     knotdata.numKnots());
-
-                if (j.value("to_disk", false)) {
-                    std::ofstream f(atoms[i].name + "-" + atoms[k].name + "_tabulated.dat"); // output file
-                    f << "# r splined exact\n";
-                    Point r = {dr, 0, 0}; // variable distance vector between particle a and b
-                    for (; r.x() < sqrt(rmax2); r.x() += dr)
-                        f << r.x() << " " << operator()(a, b, r.x() * r.x(), r) << " "
-                          << this->umatrix(i, k)(a, b, r.x() * r.x(), r) << "\n";
+    for (size_t i = 0; i < Faunus::atoms.size(); ++i) { // loop over atom types
+        for (size_t j = 0; j <= i; ++j) {               // and build matrix of spline data (knots) for each pair
+            if (atoms[i].implicit || atoms[j].implicit) {
+                continue;
+            }
+            double rmin = 0.5 * (Faunus::atoms[i].sigma + Faunus::atoms[j].sigma);
+            double rmax = js.value("rmax", rmin * 10);
+            if (auto it = js.find("cutoff_g2g"); it != js.end()) {
+                if (it->is_number()) {
+                    rmax = it->get<double>();
+                } else if (it->is_object()) {
+                    rmax = it->at("default").get<double>();
                 }
             }
+            rmin = findLowerDistance(i, j, energy_at_rmin, rmin);
+            rmax = findUpperDistance(i, j, energy_at_rmax, rmax);
+            assert(rmin < rmax);
+            createKnots(i, j, rmin, rmax);
         }
     }
+    if (js.value("to_disk", false)) {
+        save_potentials();
+    }
+}
+
+TabulatedPotential::TabulatedPotential(const std::string &name) : FunctorPotential(name) {}
+
+/**
+ * @param i Atom index
+ * @param j Atom index
+ * @param rmin Minimum splining distance
+ * @param rmax Maximum splining distance
+ */
+void TabulatedPotential::createKnots(int i, int j, double rmin, double rmax) {
+    Particle particle1 = Faunus::atoms.at(i);
+    Particle particle2 = Faunus::atoms.at(j);
```

Comment:
Indeed correct, although I think it's overkill in these small functions. `const` could be added in quite a few places but I think it clutters the otherwise pretty code.